### ISSUE TITLE
How to make links optional based on object state?

### ISSUE THREAD

**Issue by brvoisin** [jtc42/fastapi-hypermodel#8]
Hello!

I would like to have a different list of links in the response based on the state of the object. I tried with the example below.

```python
from collections import OrderedDict
from dataclasses import asdict
from dataclasses import dataclass
from enum import Enum
from typing import List
from uuid import UUID
from uuid import uuid4

from fastapi import FastAPI
from fastapi import HTTPException
from fastapi_hypermodel import HALFor
from fastapi_hypermodel import HyperModel
from fastapi_hypermodel import LinkSet
from pydantic import BaseModel


app = FastAPI()
HyperModel.init_app(app)


#### Persistence in memory ####


ITEMS = OrderedDict()


class ItemState(Enum):
    locked = "locked"
    unlocked = "unlocked"


@dataclass
class Item:
    id: UUID
    name: str
    state: ItemState = ItemState.unlocked


def _create_item(name: str):
    item = Item(id=uuid4(), name=name)
    ITEMS[str(item.id)] = item
    return item


def _read_item(item_id: str):
    return ITEMS[item_id]


def _read_all():
    return ITEMS.values()


#### Web API ####


class ItemIn(BaseModel):
    name: str


class ItemOut(HyperModel):
    id: UUID
    name: str
    state: ItemState
    links = LinkSet(
        {
            "self": HALFor(
                "get_item",
                {"item_id": "<id>"},
            ),
            # How to make link optional based on state?
            "update": HALFor(
                "update_item",
                {"item_id": "<id>"},
            ),
            "lock": HALFor(
                "lock_item",
                {"item_id": "<id>"},
            ),
        }
    )


class ItemOutLocked(ItemOut):
    links = LinkSet(
        {
            "self": HALFor(
                "get_item",
                {"item_id": "<id>"},
            ),
            "unlock": HALFor(
                "unlock_item",
                {"item_id": "<id>"},
            ),
        }
    )


@app.post("/items/", response_model=ItemOut)
async def create_item(item_in: ItemIn):
    return _create_item(item_in.name)


@app.get("/items/", response_model=List[ItemOut])
async def get_items():
    return list(_read_all())


@app.get("/items/{item_id}/", response_model=ItemOut)
async def get_item(item_id):
    item = _read_item(item_id)
    # Attempt to change the response model based on state
    if item.state == ItemState.locked:
        item = ItemOutLocked(**asdict(item))
    return item


@app.put("/items/{item_id}/", response_model=ItemOut)
async def update_item(item_id, item_in: ItemIn):
    item = _read_item(item_id)
    if item.state == ItemState.locked:
        raise HTTPException(status_code=403)
    item.name = item_in.name
    return item


@app.put("/items/{item_id}/lock", response_model=ItemOut)
async def lock_item(item_id):
    item = _read_item(item_id)
    item.state = ItemState.locked
    return item


@app.put("/items/{item_id}/unlock", response_model=ItemOut)
async def unlock_item(item_id):
    item = _read_item(item_id)
    item.state = ItemState.unlocked
    return item
```

See the response of the "get item" when it is unlocked and locked:

```bash
curl -X POST http://127.0.0.1:8000/items/ -d '{"name": "test"}' -H "Content-Type: application/json"
curl http://127.0.0.1:8000/items/085d3f3d-2566-4781-90fa-621453260452/ -s | jq .
```
```json
{
  "id": "085d3f3d-2566-4781-90fa-621453260452",
  "name": "test",
  "state": "unlocked",
  "links": {
    "self": {
      "href": "/items/085d3f3d-2566-4781-90fa-621453260452/",
      "method": "GET",
      "description": null
    },
    "update": {
      "href": "/items/085d3f3d-2566-4781-90fa-621453260452/",
      "method": "PUT",
      "description": null
    },
    "lock": {
      "href": "/items/085d3f3d-2566-4781-90fa-621453260452/lock",
      "method": "PUT",
      "description": null
    }
  }
}
```
```bash
curl -X PUT http://127.0.0.1:8000/items/085d3f3d-2566-4781-90fa-621453260452/lock
curl http://127.0.0.1:8000/items/085d3f3d-2566-4781-90fa-621453260452/ -s | jq .
```
```json
{
  "id": "085d3f3d-2566-4781-90fa-621453260452",
  "name": "test",
  "state": "locked",
  "links": {
    "self": {
      "href": "/items/085d3f3d-2566-4781-90fa-621453260452/",
      "method": "GET",
      "description": null
    },
    "unlock": {
      "href": "/items/085d3f3d-2566-4781-90fa-621453260452/unlock",
      "method": "PUT",
      "description": null
    }
  }
}
```

It is what I want, but I have to extract the following code in a function and call it each time I return a ItemOut.
```python
    if item.state == ItemState.locked:
        item = ItemOutLocked(**asdict(item))
```
In addition, the links from ItemOutLocked aren't in the documentation.

Do you know a more elegant way to do that with a correct integration in the documentation?

**Comment by jtc42** [comment]
Hmm I'll have to have a think about this. There's no nice way I know of to do this currently. It would likely involve changes to both the `LinkSet` and `HALFor`/`URLFor` classes, as the condition for each link would need to be read and parsed at runtime, which may not be trivial.

What you *could* do is create a subclass of `HALFor` that has an additional `enabled` or `active` attribute, and write a way to check the condition at runtime? At least then you wouldn't also need to modify the `LinkSet` class.

This is really just a brain-dump for now, there may be a super elegant solution to this I'm just not aware of, so open to suggestions and PRs.

**Comment by tmuguet** [comment]
Hi,

Coming late here, but I had the same question and found a not so inelegant solution.

First, extend classes `UrlFor` and `HALFor`, which take a filter callback at init, and execute that filter at runtime:

```python
class UrlForConditional(UrlFor):
    _condition: Callable[[Dict[str, Any]], bool]

    def __init__(
        self,
        condition: Callable[[Dict[str, Any]], bool],
        endpoint: str,
        param_values: Optional[Dict[str, str]] = None,
    ) -> None:
        super().__init__(endpoint, param_values)
        self._condition = condition  # type:ignore

    def __build_hypermedia__(
        self, app: Optional[FastAPI], values: Dict[str, Any]
    ) -> Optional[str]:
        if not self._condition(values):  # type:ignore
            return None
        return super().__build_hypermedia__(app, values)


class ConditionalHALFor(HALFor):
    _condition: Callable[[Dict[str, Any]], bool] = PrivateAttr()

    def __init__(
        self,
        condition: Callable[[Dict[str, Any]], bool],
        endpoint: str,
        param_values: Optional[Dict[str, str]] = None,
        description: Optional[str] = None,
    ) -> None:
        super().__init__(endpoint, param_values, description)
        self._condition = condition  # type:ignore

    def __build_hypermedia__(
        self, app: Optional[FastAPI], values: Dict[str, Any]
    ) -> Optional[HALItem]:
        if not self._condition(values):  # type:ignore
            return None
        return super().__build_hypermedia__(app, values)
```

Then to use them:

```python
class ItemOut(HyperModel):
    id: UUID
    name: str
    state: ItemState
    links = LinkSet(
        {
            "self": HALFor(
                "get_item",
                {"item_id": "<id>"},
            ),
            "update": ConditionalHALFor(
                lambda values: values['state'] != ItemState.locked,
                "update_item",
                {"item_id": "<id>"},
            ),
            "lock": ConditionalHALFor(
                lambda values: values['state'] != ItemState.locked,
                "lock_item",
                {"item_id": "<id>"},
            ),
            "unlock": ConditionalHALFor(
                lambda values: values['state'] == ItemState.locked,
                "unlock_item",
                {"item_id": "<id>"},
            ),
        }
    )
```

This will make *all* URLs appear in the documentation.

It will generate empty URLs if condition is unmet though. I ended up also extending `LinkSet` to discard any URL that is `None`:

```python
class ConditionalLinkSet(LinkSet):
    def __build_hypermedia__(
        self, app: Optional[FastAPI], values: Dict[str, Any]
    ) -> Dict[str, str]:
        links = super().__build_hypermedia__(app, values)
        return {k: u for k, u in links.items() if u is not None}
```

**Comment by jtc42** [comment]
Big thanks to @tmuguet, this has now been released in v0.4.0
